File: quartz/plugins/transformers/mathblock-fixer.ts
import { QuartzTransformerPlugin } from "../types"
import fs from "fs"
import path from "path"

export interface Options {
  debug: boolean
}

const defaultOptions: Options = {
  debug: false,
}

export const MathBlockFixer: QuartzTransformerPlugin<Partial<Options>> = (userOpts) => {
  const opts = { ...defaultOptions, ...userOpts }

  const fixMathBlocks = (content: string): string => {
    const mathBlockPattern = /\$\$.*?\$\$/gs
    const endDollarPattern = /^\s*>\s*\$\$$|^\s*\$\$$/g

    content = content.replace(/(\d+\.)\s?\$\$/g, '$1\n$$')

    const fixBlock = (block: string) => {
      let lines = block.split('\n')
      // Ensure the first line is just '$$'
      if (lines[0].trim() !== '$$') {
        lines[0] = '$$'
        if (lines[1]) {
          lines.splice(1, 0, lines[1].trim().slice(2))
        }
      }
      // Ensure the last line is just '$$'
      const lastLine = lines[lines.length - 1].trim()
      if (lastLine !== '$$') {
        lines[lines.length - 1] = lines[lines.length - 1].trim().slice(0, -2)
        lines.push('$$')
      }
      return lines.join('\n')
    }

    content = content.replace(mathBlockPattern, (match) => fixBlock(match))

    content = content.replace(/    /g, '\t')
    content = content.replace(/ \t/g, '\t')

    let lines = content.split('\n')
    let insideCallout = false
    let insideTabBlock = false
    let insideDoubleTabBlock = false
    let insideTabbedCallout = false
    let insideTabInCallout = false
    let insideDoubleTabBlockInCallout = false

    for (let i = 0; i < lines.length; i++) {
      let line = lines[i]

      // Handle double tabs inside callout
      if (line.startsWith('>\t\t')) {
        insideDoubleTabBlockInCallout = true
      } else if (line.trim() === '' || line.trim().startsWith('#') || line.trim().startsWith('---') || line.trim().startsWith('```')) {
        insideDoubleTabBlockInCallout = false
      }
      if (insideDoubleTabBlockInCallout && !line.startsWith('>\t\t') && !/^\>\t\d+\./.test(line.trim()) && !/^\>\t\- /.test(line.trim())) {
        if (line.trim().startsWith('>\t')) {
          lines[i] = '>\t\t' + line.trim().slice(2)
        } else if (line.trim().startsWith('>')) {
          lines[i] = '>\t\t' + line.trim().slice(1)
        } else {
          lines[i] = '>\t\t' + line
        }
        continue
      }

      // Handle double tabs
      if (line.startsWith('\t\t')) {
        insideDoubleTabBlock = true
      } else if (line.trim() === '' || line.trim().startsWith('#') || line.trim().startsWith('---') || line.trim().startsWith('```')) {
        insideDoubleTabBlock = false
      }
      if (insideDoubleTabBlock && !line.startsWith('\t\t') && !/^\d+\./.test(line.trim()) && !/^\t\- /.test(line.trim())) {
        if (line.startsWith('\t')) {
          lines[i] = '\t' + line
        } else {
          lines[i] = '\t\t' + line
        }
        continue
      }

      // Handle tabbed callouts
      if (/^\t ?\>/.test(line)) {
        insideTabbedCallout = true
      } else if (line.trim() === '' || line.trim().startsWith('#') || line.trim().startsWith('---') || line.trim().startsWith('```')) {
        insideTabbedCallout = false
      }
      if (!insideTabbedCallout && /^\t ?\>/.test(line.trim())) {
        lines[i] = line.trim().slice(1)
      } else if (insideTabbedCallout && !line.startsWith('\t>')) {
        if (line.trim().startsWith('\t')) {
          lines[i] = '>' + line
        } else if (line.trim().startsWith('>')) {
          lines[i] = '\t' + line
        } else {
          lines[i] = '\t>' + line
        }
        continue
      }

      // Handle tabs in callouts
      if (/^\> ?\t/.test(line) || /^\> ?\d+\./.test(line.trim())) {
        insideTabInCallout = true
      } else if (line.trim() === '' || line.trim().startsWith('#') || line.trim().startsWith('---') || line.trim().startsWith('```')) {
        insideTabInCallout = false
      }

      if (insideTabInCallout && /^\> ?\t ?/.test(line)) {
        lines[i] = line.replace(/^\> ?\t ?/, '>\t')
        continue
      }

      if (insideTabInCallout && !/^\> ?\t/.test(line.trim()) && !/^\> ?\t/.test(line) && !/^\> ?\d+\./.test(line.trim()) && !/^\- /.test(line.trim())) {
        if (/^\d+\./.test(line.trim())) {
          lines[i] = '>' + line
        } else if (line.trim().startsWith('>')) {
          lines[i] = '>\t' + line.trim().slice(1)
        } else {
          lines[i] = '>\t' + line
        }
        continue
      }

      // Handle callouts
      if (line.trim().startsWith('>')) {
        insideCallout = true
      } else if (line.trim() === '' || line.trim().startsWith('#') || line.trim().startsWith('---') || line.trim().startsWith('```')) {
        insideCallout = false
      }
      if (insideCallout && !line.trim().startsWith('>')) {
        lines[i] = '> ' + line
        continue
      }

      // Handle tabs
      if (/^\d+\./.test(line.trim()) || /^\- /.test(line.trim())) {
        insideTabBlock = true
      } else if (line.trim() === '' || line.trim().startsWith('#') || line.trim().startsWith('---') || line.trim().startsWith('```')) {
        insideTabBlock = false
      }
      if (!insideTabBlock && /^ ?\t/.test(line)) {
        lines[i] = line.trim().slice(1)
      } else if (insideTabBlock && !/^ ?\t/.test(line) && !/^\d+\./.test(line.trim()) && !/^\- /.test(line.trim())) {
        lines[i] = '\t' + line
        continue
      }
    }

    return lines.join('\n')
  }

  const writeDebugFile = (content: string, filePath: string) => {
    const debugDir = path.resolve('mathblock-debug')
    if (!fs.existsSync(debugDir)) {
      fs.mkdirSync(debugDir)
    }
    const debugFilePath = path.join(debugDir, path.basename(filePath))
    fs.writeFileSync(debugFilePath, content, 'utf-8')
  }

  return {
    name: "MathBlockFixer",
    textTransform(_ctx, src) {
      if (src.includes('[[#סמסטר א’|קפוץ לחומר]].')) {
        return src
      }

      const content = fixMathBlocks(src.toString())
      if (opts.debug) {
        // TODO: Find a way to get the file path, currently generating random file names
        const random = Math.random().toString(36).substring(7)
        const filePath = random + '.md'
        writeDebugFile(content, filePath)
      }
      return content
    },
  }
}

File: scripts/mathblock-fixer.py
# For a given file (for testing, let's use PHY2\PHY2_002 אלקטרוסטטיקה - שטף חשמלי, חוק גאוס ופוטנציאל חשמלי.md), this script will fix the math blocks in the file.
# For each display math block (i.e. a block that starts with $$ and ends with $$), it will:
# 1. Put the beginning $$ and ending $$ on a separate line.
# 2. If the math block is in a callout (i.e., begins with >), then ensure that each new line in the math block is also prefixed with >.

import re

def fix_math_blocks(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        content = file.read()

    # Regex to find all display math blocks
    math_block_pattern = re.compile(r'(\$\$.*?\$\$)', re.DOTALL)
    end_dollar_pattern = re.compile(r'^\s*>\s*\$\$$|^\s*\$\$$')

    # Look for numbers followed by '$$' on the same line, and separate them into two lines
    content = re.sub(r'(\d+\.)\s?\$\$', r'\1\n$$', content)

    def fix_block(block):
        # Ensure $$ are on separate lines if not already
        lines = block.split('\n')
        if not lines[0].strip() == '$$':
            lines[0] = '$$\n' + lines[0].lstrip('$')
        if not end_dollar_pattern.match(lines[-1].strip()):
            lines[-1] = lines[-1].rstrip('$') + '\n$$'
        return '\n'.join(lines)

    # Replace each math block with the fixed version
    content = math_block_pattern.sub(lambda m: fix_block(m.group(1)), content)

    # Replace all 4 spaces with tabs
    content = content.replace('    ', '\t')
    content = content.replace(' \t', '\t')

    # Split content into lines and process each line for callouts and tabs
    lines = content.split('\n')
    inside_callout = False
    inside_tab_block = False
    inside_double_tab_block = False
    inside_tabbed_callout = False
    inside_tab_in_callout = False
    inside_double_tab_block_in_callout = False
    for i, line in enumerate(lines):

        # Handle double tabs inside callout
        if line.startswith('>\t\t'):
            inside_double_tab_block_in_callout = True
        elif line.strip() == '' or line.strip().startswith('#') or line.strip().startswith('---'):
            inside_double_tab_block_in_callout = False
        if inside_double_tab_block_in_callout and not line.startswith('>\t\t') and not re.match(r'^>\t\d+\.', line.strip()) and not re.match(r'^>\t\- ', line.strip()):
            if line.strip().startswith('>\t'):
                lines[i] = '>\t\t' + line.strip().lstrip('>\t')
            elif line.strip().startswith('>'):
                lines[i] = '>\t\t' + line.strip().lstrip('>')
            else:
                lines[i] = '>\t\t' + line
            continue
        
        # Handle double tabs
        if line.startswith('\t\t'):
            inside_double_tab_block = True
        elif line.strip() == '' or line.strip().startswith('#') or line.strip().startswith('---'):
            inside_double_tab_block = False
        if inside_double_tab_block and not line.startswith('\t\t') and not re.match(r'^\d+\.', line.strip()) and not re.match(r'^\t\- ', line.strip()):
            if line.startswith('\t'):
                lines[i] = '\t' + line
            else:
                lines[i] = '\t\t' + line
            continue
            

        # Handle tabbed callouts
        if re.match(r'^\t ?\>', line):
            inside_tabbed_callout = True
        elif line.strip() == '' or line.strip().startswith('#') or line.strip().startswith('---'):
            inside_tabbed_callout = False
        if not inside_tabbed_callout and re.match(r'^\t ?\>', line.strip()):
            lines[i] = line.strip().lstrip('\t>')
        elif inside_tabbed_callout and not line.startswith('\t>'):
            if line.strip().startswith('\t'):
                lines[i] = '>' + line
            elif line.strip().startswith('>'):
                lines[i] = '\t' + line
            else:
                lines[i] = '\t>' + line
            continue
            
            

        # Handle tabs in callouts
        if re.match(r'^\> ?\t', line) or re.match(r'^\> ?\d+\.', line.strip()):
            inside_tab_in_callout = True
        elif line.strip() == '' or line.strip().startswith('#') or line.strip().startswith('---'):
            inside_tab_in_callout = False

        if inside_tab_in_callout and re.match(r'^\> ?\t ?', line):
            lines[i] = re.sub(r'^\> ?\t ?', '>\t', line)
            continue

        if inside_tab_in_callout and not re.match(r'^\> ?\t', line.strip()) and not re.match(r'^\> ?\t', line) and not re.match(r'^\> ?\d+\.', line.strip()) and not re.match(r'^\- ', line):
            if re.match(r'\d+\.', line.strip()):
                lines[i] = '>' + line
            elif line.strip().startswith('>'):
                # Replace '>' with '>\t'
                lines[i] = '>\t' + line.strip().lstrip('>')
            else:
                lines[i] = '>\t' + line
            continue
            

        # Handle callouts
        if line.strip().startswith('>'):
            inside_callout = True
        elif line.strip() == '' or line.strip().startswith('#') or line.strip().startswith('---'):
            inside_callout = False
        if inside_callout and not line.strip().startswith('>'):
            lines[i] = '> ' + line
            continue


        # Handle tabs
        if re.match(r'^\d+\.', line.strip()) or re.match(r'^\- ', line.strip()):
            inside_tab_block = True
        elif line.strip() == '' or line.strip().startswith('#') or line.strip().startswith('---'):
            inside_tab_block = False
        if not inside_tab_block and re.match(r'^ ?\t', line):
            lines[i] = line.strip().lstrip('\t')
        elif inside_tab_block and not re.match(r'^ ?\t', line) and not re.match(r'^\d+\.', line.strip()) and not re.match(r'^\- ', line.strip()):
            lines[i] = '\t' + line
            continue


    # Join lines back into content
    fixed_content = '\n'.join(lines)

    with open(file_path, 'w', encoding='utf-8') as file:
        file.write(fixed_content)

# Run the script on all the .md files in the 'content' directory and below it, but exclude '.excalidraw.md' files
import os
content_dir = os.path.abspath('content/Technion')
for root, dirs, files in os.walk(content_dir):
    for file in files:
        if file.endswith('.md') and not file.endswith('.excalidraw.md'):
            fix_math_blocks(os.path.join(root, file))
            print(f'Fixed math blocks in {os.path.join(root, file)}')


# Run the script on 'content/mathblock-fixer-test.md' for testing
# fix_math_blocks('content/test/mathblock-fixer-test.md')
# print('Fixed math blocks in content/test/mathblock-fixer-test.md')